File: web-greeter/greeter.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
#
#  greeter.py
#
#  Copyright © 2017 Antergos
#
#  This file is part of Web Greeter.
#
#  Web Greeter is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 3 of the License, or
#  (at your option) any later version.
#
#  Web Greeter is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  The following additional terms are in effect as per Section 7 of the license:
#
#  The preservation of all legal notices and author attributions in
#  the material or in the Appropriate Legal Notices displayed
#  by works containing it is required.
#
#  You should have received a copy of the GNU General Public License
#  along with Web Greeter; If not, see <http://www.gnu.org/licenses/>.

# Standard Lib
import sys
import yaml
import pkg_resources
import os
from typing import (
    ClassVar,
    Type,
    List,
    Tuple,
)
from logging import (
    getLogger,
    DEBUG,
    Formatter,
    StreamHandler,
)

# 3rd-Party Libs

# This Application
from utils import config

import globals
from globals import WebGreeter

def loadWhitherConf():
    global whither_yaml
    global webGreeter_conf
    global file_test
    try:
        file_test = pkg_resources.resource_string("__main__", 'whither.yml').decode('utf-8')
    except Exception:
        file_test = pkg_resources.resource_string(__file__, 'whither.yml').decode('utf-8')

    whither_yaml = yaml.safe_load(file_test)
    webGreeter_conf = whither_yaml["WebGreeter"]


def debugLog(txt: str):
    log_format = ''.join([
        '%(asctime)s [ %(levelname)s ] %(module)s - %(filename)s:%(',
        'lineno)d : %(funcName)s | %(message)s'
    ])

    formatter = Formatter(fmt=log_format, datefmt="%Y-%m-%d %H:%M:%S")
    stream_handler = StreamHandler()
    logger = getLogger()

    stream_handler.setLevel(DEBUG)
    stream_handler.setFormatter(formatter)
    logger.setLevel(DEBUG)
    logger.addHandler(stream_handler)
    logger.debug(txt)

def show_help():
    version = webGreeter_conf["app"]["version"]["full"]
    help_text = """Usage:
  web-greeter [OPTION...] - LightDM Web Greeter

  --debug                   Runs the greeter in debug mode
  --normal                  Runs in non-debug mode

  --list                    Lists available themes
  --theme                   Sets the theme to use

  -h, --help                Show this help list
  -v, --version             Print program version""".format(
    version = version
)
    print(help_text)


def show_version():
    version = webGreeter_conf["app"]["version"]["full"]
    print("{version}".format(version = version))

def changeConfig(option: str, value):
    custom_config[option] = value
    return

def debugMode(value: bool):
    if value:
        custom_config["debug_mode"] = True
        custom_config["decorated"] = True
        custom_config["stays_on_top"] = False
    else:
        custom_config["debug_mode"] = False
        custom_config["decorated"] = False
        custom_config["stays_on_top"] = True
    pass

def changeTheme(theme: str):
    dirlist = listThemes(True)

    if theme in dirlist:
        custom_config["theme"] = theme
    else:
        debugLog("Theme not found. Going with config theme")
    return

def listThemes(quiet = False):
    themes_dir = webGreeter_conf["app"]["themes_dir"]
    themes_dir = themes_dir if os.path.exists(themes_dir) else "/usr/share/web-greeter/themes"
    filenames = os.listdir(themes_dir)

    dirlist = []
    for file in filenames:
        if os.path.isdir(os.path.join(themes_dir, file)):
            dirlist.append(file)

    if not quiet:
        print("Themes are located in {themes_dir}\n".format(themes_dir = themes_dir))
        for theme in dirlist:
            print("-", theme)

    return dirlist

args_lenght = sys.argv.__len__()

def yargs(args: List[str]):
    loadWhitherConf()
    used = 0

    if args[0] == "--help" or args[0] == "-h":
        show_help()
        used+=1
        exit()
    elif args[0] == "--version" or args[0] == "-v":
        show_version()
        used+=1
        exit()
    elif args[0] == "--debug":
        changeConfig("debug_mode", True)
        used+=1
        pass
    elif args[0] == "--normal":
        changeConfig("debug_mode", False)
        used+=1
        pass
    elif args[0] == "--theme":
        if args.__len__() > 1 :
            changeTheme(args[1])
            used+=2
        else:
            print("No theme provided")
            used+=1
            exit(1)
    elif args[0] == "--list":
        listThemes()
        used+=1
        exit()
    else:
        show_help()
        used+=1
        exit(1)
    for x in range(used):
        args.pop(0)
    if args.__len__() != 0:
        yargs(args)
    pass

if __name__ == '__main__':
    custom_config = globals.custom_config

    if args_lenght > 1:
        args = sys.argv
        args.pop(0)
        yargs(args)


    globals.greeter = WebGreeter()

    globals.greeter.run()

File: web-greeter/resources/js/docs/Greeter.js
/*
 * LightDMGreeter.js
 *
 * Copyright © 2017 Antergos Developers <dev@antergos.com>
 *
 * This file is part of Web Greeter.
 *
 * Web Greeter is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * Web Greeter is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * The following additional terms are in effect as per Section 7 of the license:
 *
 * The preservation of all legal notices and author attributions in
 * the material or in the Appropriate Legal Notices displayed
 * by works containing it is required.
 *
 * You should have received a copy of the GNU General Public License
 * along with web-greeter; If not, see <http://www.gnu.org/licenses/>.
 */


/**
 * Base class for the greeter's Theme JavaScript API. Greeter themes will interact
 * directly with an object derived from this class to facilitate the user log-in process.
 * The greeter will automatically create an instance when it starts.
 * The instance can be accessed using the global variable: [`lightdm`](#dl-window-lightdm).
 *
 * @memberOf LightDM
 */
class Greeter {

	constructor() {
		if ( 'lightdm' in window ) {
			return window.lightdm;
		}

		window.lightdm = GreeterUtils.bind_this( this );

		return window.lightdm;
	}

	/**
	 * The username of the user being authenticated or {@link null}
	 * if there is no authentication in progress.
	 * @type {string|null}
	 * @readonly
	 */
	get authentication_user() {}

	/**
	 * Whether or not the guest account should be automatically logged
	 * into when the timer expires.
	 * @type {boolean}
	 * @readonly
	 */
	get autologin_guest() {}

	/**
	 * The number of seconds to wait before automatically logging in.
	 * @type {number}
	 * @readonly
	 */
	get autologin_timeout() {}

	/**
	 * The username with which to automatically log in when the timer expires.
	 * @type {string}
	 * @readonly
	 */
	get autologin_user() {}

	/**
	 * Whether or not the greeter can make the system hibernate.
	 * @type {boolean}
	 * @readonly
	 */
	get can_hibernate() {}

	/**
	 * Whether or not the greeter can make the system restart.
	 * @type {boolean}
	 * @readonly
	 */
	get can_restart() {}

	/**
	 * Whether or not the greeter can make the system shutdown.
	 * @type {boolean}
	 * @readonly
	 */
	get can_shutdown() {}

	/**
	 * Whether or not the greeter can make the system suspend/sleep.
	 * @type {boolean}
	 * @readonly
	 */
	get can_suspend() {}

	/**
	 * The name of the default session.
	 * @type {string}
	 * @readonly
	 */
	get default_session() {}

	/**
	 * Whether or not guest sessions are supported.
	 * @type {boolean}
	 * @readonly
	 */
	get has_guest_account() {}

	/**
	 * Whether or not user accounts should be hidden.
	 * @type {boolean}
	 * @readonly
	 */
	get hide_users_hint() {}

	/**
	 * The system's hostname.
	 * @type {string}
	 * @readonly
	 */
	get hostname() {}

	/**
	 * Whether or not the greeter is in the process of authenticating.
	 * @type {boolean}
	 * @readonly
	 */
	get in_authentication() {}

	/**
	 * Whether or not the greeter has successfully authenticated.
	 * @type {boolean}
	 * @readonly
	 */
	get is_authenticated() {}

	/**
	 * The current language or {@link null} if no language.
	 * @type {LightDM.Language|null}
	 * @readonly
	 */
	get language() {}

	/**
	 * A list of languages to present to the user.
	 * @type {LightDM.Language[]}
	 * @readonly
	 */
	get languages() {}

	/**
	 * The currently active layout for the selected user.
	 * @type {LightDM.Layout}
	 */
	get layout() {}

	/**
	 * Set the active layout for the selected user.
	 * @param {LightDM.Layout} value
	 */
	set layout( value ) {}

	/**
	 * A list of keyboard layouts to present to the user.
	 * @type {LightDM.Layout[]}
	 * @readonly
	 */
	get layouts() {}

	/**
	 * Whether or not the greeter was started as a lock screen.
	 * @type {boolean}
	 * @readonly
	 */
	get lock_hint() {}

	/**
	 * The available remote sessions.
	 * @type {LightDM.Session[]}
	 * @readonly
	 */
	get remote_sessions() {}

	/**
	 * Whether or not the guest account should be selected by default.
	 * @type {boolean}
	 * @readonly
	 */
	get select_guest_hint() {}

	/**
	 * The username to select by default.
	 * @type {string}
	 * @readonly
	 */
	get select_user_hint() {}

	/**
	 * List of available sessions.
	 * @type {LightDM.Session[]}
	 * @readonly
	 */
	get sessions() {}

	/**
	 * Check if a manual login option should be shown. If {@link true}, the theme should
	 * provide a way for a username to be entered manually. Otherwise, themes that show
	 * a user list may limit logins to only those users.
	 * @type {string}
	 * @readonly
	 */
	get show_manual_login_hint() {}

	/**
	 * Check if a remote login option should be shown. If {@link true}, the theme should provide
	 * a way for a user to log into a remote desktop server.
	 * @type {string}
	 * @readonly
	 * @internal
	 */
	get show_remote_login_hint() {}

	/**
	 * List of available users.
	 * @type {LightDM.User[]}
	 * @readonly
	 */
	get users() {}


	/**
	 * Starts the authentication procedure for a user.
	 *
	 * @arg {String|null} username A username or {@link null} to prompt for a username.
	 */
	authenticate( username = null ) {}

	/**
	 * Starts the authentication procedure for the guest user.
	 */
	authenticate_as_guest() {}

	/**
	 * Cancel the user authentication that is currently in progress.
	 */
	cancel_authentication() {}

	/**
	 * Cancel the automatic login.
	 */
	cancel_autologin() {}

	/**
	 * Triggers the system to hibernate.
	 * @returns {boolean} {@link true} if hibernation initiated, otherwise {@link false}
	 */
	hibernate() {}

	/**
	 * Provide a response to a prompt.
	 * @arg {string} response
	 */
	respond( response ) {}

	/**
	 * Triggers the system to restart.
	 * @returns {boolean} {@link true} if restart initiated, otherwise {@link false}
	 */
	restart() {}

	/**
	 * Set the language for the currently authenticated user.
	 * @arg {string} language The language in the form of a locale specification (e.g. 'de_DE.UTF-8')
	 * @returns {boolean} {@link true} if successful, otherwise {@link false}
	 */
	set_language( language ) {}

	/**
	 * Triggers the system to shutdown.
	 * @returns {boolean} {@link true} if shutdown initiated, otherwise {@link false}
	 */
	shutdown() {}

	/**
	 * Start a session for the authenticated user.
	 * @arg {String|null} session The session to log into or {@link null} to use the default.
	 * @returns {boolean} {@link true} if successful, otherwise {@link false}
	 */
	start_session( session ) {}

	/**
	 * Triggers the system to suspend/sleep.
	 * @returns {boolean} {@link true} if suspend/sleep initiated, otherwise {@link false}
	 */
	suspend() {}

	/**
	 * Gets the brightness
	 * @type {Number}
	 * @readonly
	 */
	get brightness() {}

	/**
	 * Set the brightness to quantity
	 * @arg {Number} quantity The quantity to set
	 */
	brightnessSet( quantity ) {}

	/**
	 * Increase the brightness by quantity
	 * @arg {Number} quantity The quantity to increase
	 */
	brightnessIncrease( quantity ) {}

	/**
	 * Decrease the brightness by quantity
	 * @arg {Number} quantity The quantity to decrease
	 */
	brightnessDecrease( quantity ) {}

	/**
	 * Gets the battery data
	 * @type {Object}
	 * @readonly
	 */
	get batteryData() {}

	/**
	 * Whether or not the greeter can access to battery data
	 * @type {boolean}
	 * @readonly
	 */
	get can_access_battery() {}

	/**
	 * Updates the battery data
	 */
	batteryUpdate() {}

	/**
	 * Whether or not the greeter can control display brightness
	 * @type {boolean}
	 * @readonly
	 */
	get can_access_brightness() {}

}


/**
 * Moment.js instance - Loaded and instantiated automatically by the greeter.
 * @name moment
 * @type {object}
 * @version 2.17.0
 * @memberOf window
 * @see [Moment.js Documentation](http://momentjs.com/docs)
 */

/**
 * jQuery instance - Themes must manually load the included vendor script in order to use this object.
 * @name jQuery
 * @type {object}
 * @version 3.1.1
 * @memberOf window
 * @see [jQuery Documentation](http://api.jquery.com)
 */

/**
 * jQuery instance
 * @name $
 * @memberOf window
 * @see {@link window.jQuery}
 */

/**
 * JS-Cookie instance - Themes must manually load the included vendor script in order to use this object.
 * @name Cookies
 * @type {object}
 * @version 2.1.3
 * @memberOf window
 * @see [JS Cookie Documentation](https://github.com/js-cookie/js-cookie/tree/latest#readme)
 */



